osu.Game.Rulesets.Taiko/Replays/TaikoAutoGenerator.cs

```diff
@@ -113,14 +113,21 @@ public override Replay Generate()
                 else
                     throw new InvalidOperationException("Unknown hit object type.");
 
-                Frames.Add(new TaikoReplayFrame(endTime + KEY_UP_DELAY));
-
                 if (i < Beatmap.HitObjects.Count - 1)
                 {
-                    double waitTime = Beatmap.HitObjects[i + 1].StartTime - 1000;
+                    var nextHitObject = Beatmap.HitObjects[i + 1];
+
+                    if (nextHitObject.StartTime > endTime + KEY_UP_DELAY)
```

Comment:
`!(nextHitObject.StartTime < endTime + KEY_UP_DELAY)` is inverted to `nextHitObject.StartTime >= endTime + KEY_UP_DELAY` but you're checking if greater than, was that intended?

Reply:
I did change that to ``nextHitObject.StartTime >= endTime + KEY_UP_DELAY`` in my latest commit. And yes, it is intended. The logic is: if the next press happens later than 50ms or if it's the last note, perform key up frame after 50ms. Otherwise, if the next press is required in less than 50ms - don't add any frames. The next key down frame will work as key up & key down, so there is no need to add key up frame.